[WebGPU] Safe C++ adaption of APIConversions.h

mwyrzykowski opened the pull request:
#### 48ab5be919c779b4eec88a79c3fc2703c4e17346
<pre>
[WebGPU] Safe C++ adaption of APIConversions.h
<a href="https://bugs.webkit.org/show_bug.cgi?id=281888">https://bugs.webkit.org/show_bug.cgi?id=281888</a>
<a href="https://rdar.apple.com/138358898">rdar://138358898</a>

Reviewed by Geoffrey Garen.

Static analyzer indicated APIConversions resulted in warnings, correct
them all by switching the return type of fromAPI to Ref&lt;T&gt; from T&amp;.

* Source/WebGPU/WebGPU/APIConversions.h:
(WebGPU::fromAPI):
* Source/WebGPU/WebGPU/Adapter.mm:
(wgpuAdapterReference):
(wgpuAdapterRelease):
(wgpuAdapterEnumerateFeatures):
(wgpuAdapterGetLimits):
(wgpuAdapterGetProperties):
(wgpuAdapterHasFeature):
(wgpuAdapterRequestDevice):
(wgpuAdapterRequestDeviceWithBlock):
(wgpuAdapterXRCompatible):
* Source/WebGPU/WebGPU/BindGroup.mm:
(WebGPU::makeBindGroupEntryUsageData):
(WebGPU::Device::createBindGroup):
(wgpuBindGroupReference):
(wgpuBindGroupRelease):
(wgpuBindGroupSetLabel):
(wgpuBindGroupUpdateExternalTextures):
* Source/WebGPU/WebGPU/BindGroupLayout.mm:
(wgpuBindGroupLayoutReference):
(wgpuBindGroupLayoutRelease):
(wgpuBindGroupLayoutSetLabel):
* Source/WebGPU/WebGPU/Buffer.mm:
(wgpuBufferReference):
(wgpuBufferRelease):
(wgpuBufferDestroy):
(wgpuBufferGetMapState):
(wgpuBufferGetMappedRange):
(wgpuBufferGetBufferContents):
(wgpuBufferGetInitialSize):
(wgpuBufferGetCurrentSize):
(wgpuBufferMapAsync):
(wgpuBufferMapAsyncWithBlock):
(wgpuBufferUnmap):
(wgpuBufferSetLabel):
(wgpuBufferGetUsage):
(wgpuBufferCopy):
* Source/WebGPU/WebGPU/CommandBuffer.mm:
(wgpuCommandBufferReference):
(wgpuCommandBufferRelease):
(wgpuCommandBufferSetLabel):
* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::errorValidatingTimestampWrites):
(WebGPU::CommandEncoder::errorValidatingRenderPassDescriptor const):
(WebGPU::CommandEncoder::beginRenderPass):
(WebGPU::CommandEncoder::errorValidatingImageCopyBuffer const):
(WebGPU::CommandEncoder::errorValidatingCopyBufferToTexture const):
(WebGPU::CommandEncoder::copyBufferToTexture):
(WebGPU::CommandEncoder::errorValidatingCopyTextureToBuffer const):
(WebGPU::CommandEncoder::clearTextureIfNeeded):
(WebGPU::CommandEncoder::copyTextureToBuffer):
(WebGPU::errorValidatingCopyTextureToTexture):
(WebGPU::CommandEncoder::copyTextureToTexture):
(wgpuCommandEncoderReference):
(wgpuCommandEncoderRelease):
(wgpuCommandEncoderBeginComputePass):
(wgpuCommandEncoderBeginRenderPass):
(wgpuCommandEncoderCopyBufferToBuffer):
(wgpuCommandEncoderCopyBufferToTexture):
(wgpuCommandEncoderCopyTextureToBuffer):
(wgpuCommandEncoderCopyTextureToTexture):
(wgpuCommandEncoderClearBuffer):
(wgpuCommandEncoderFinish):
(wgpuCommandEncoderInsertDebugMarker):
(wgpuCommandEncoderPopDebugGroup):
(wgpuCommandEncoderPushDebugGroup):
(wgpuCommandEncoderResolveQuerySet):
(wgpuCommandEncoderWriteTimestamp):
(wgpuCommandEncoderSetLabel):
* Source/WebGPU/WebGPU/ComputePassEncoder.mm:
(wgpuComputePassEncoderReference):
(wgpuComputePassEncoderRelease):
(wgpuComputePassEncoderDispatchWorkgroups):
(wgpuComputePassEncoderDispatchWorkgroupsIndirect):
(wgpuComputePassEncoderEnd):
(wgpuComputePassEncoderInsertDebugMarker):
(wgpuComputePassEncoderPopDebugGroup):
(wgpuComputePassEncoderPushDebugGroup):
(wgpuComputePassEncoderSetBindGroup):
(wgpuComputePassEncoderSetPipeline):
(wgpuComputePassEncoderSetLabel):
* Source/WebGPU/WebGPU/ComputePipeline.mm:
(wgpuComputePipelineReference):
(wgpuComputePipelineRelease):
(wgpuComputePipelineGetBindGroupLayout):
(wgpuComputePipelineSetLabel):
* Source/WebGPU/WebGPU/Device.mm:
* Source/WebGPU/WebGPU/ExternalTexture.mm:
(wgpuExternalTextureReference):
(wgpuExternalTextureRelease):
(wgpuExternalTextureDestroy):
(wgpuExternalTextureUndestroy):
(wgpuExternalTextureUpdate):
* Source/WebGPU/WebGPU/Instance.mm:
(wgpuInstanceReference):
(wgpuInstanceRelease):
(wgpuInstanceCreateSurface):
(wgpuInstanceProcessEvents):
(wgpuInstanceRequestAdapter):
(wgpuInstanceRequestAdapterWithBlock):
(wgpuBufferIsValid):
(wgpuAdapterIsValid):
(wgpuBindGroupIsValid):
(wgpuBindGroupLayoutIsValid):
(wgpuCommandBufferIsValid):
(wgpuCommandEncoderIsValid):
(wgpuComputePassEncoderIsValid):
(wgpuComputePipelineIsValid):
(wgpuDeviceIsValid):
(wgpuExternalTextureIsValid):
(wgpuPipelineLayoutIsValid):
(wgpuPresentationContextIsValid):
(wgpuQuerySetIsValid):
(wgpuQueueIsValid):
(wgpuRenderBundleEncoderIsValid):
(wgpuRenderBundleIsValid):
(wgpuRenderPassEncoderIsValid):
(wgpuRenderPipelineIsValid):
(wgpuSamplerIsValid):
(wgpuShaderModuleIsValid):
(wgpuTextureIsValid):
(wgpuTextureViewIsValid):
(wgpuXRBindingIsValid):
(wgpuXRSubImageIsValid):
(wgpuXRProjectionLayerIsValid):
(wgpuXRViewIsValid):
* Source/WebGPU/WebGPU/IsValidToUseWith.h:
(WebGPU::isValidToUseWith):
* Source/WebGPU/WebGPU/PipelineLayout.mm:
(wgpuPipelineLayoutReference):
(wgpuPipelineLayoutRelease):
(wgpuPipelineLayoutSetLabel):
* Source/WebGPU/WebGPU/PresentationContext.mm:
(wgpuSurfaceReference):
(wgpuSurfaceRelease):
(wgpuSwapChainReference):
(wgpuSwapChainRelease):
(wgpuSurfaceGetPreferredFormat):
(wgpuSwapChainGetCurrentTexture):
(wgpuSwapChainGetCurrentTextureView):
(wgpuSwapChainPresent):
(wgpuSwapChainGetTextureAsNativeImage):
* Source/WebGPU/WebGPU/QuerySet.mm:
(wgpuQuerySetReference):
(wgpuQuerySetRelease):
(wgpuQuerySetDestroy):
(wgpuQuerySetSetLabel):
(wgpuQuerySetGetCount):
(wgpuQuerySetGetType):
* Source/WebGPU/WebGPU/Queue.mm:
(WebGPU::Queue::clearTextureIfNeeded):
(WebGPU::Queue::writeTexture):
(wgpuQueueReference):
(wgpuQueueRelease):
(wgpuQueueOnSubmittedWorkDone):
(wgpuQueueOnSubmittedWorkDoneWithBlock):
(wgpuQueueSubmit):
(wgpuQueueWriteBuffer):
(wgpuQueueWriteTexture):
(wgpuQueueSetLabel):
* Source/WebGPU/WebGPU/RenderBundle.mm:
(wgpuRenderBundleReference):
(wgpuRenderBundleRelease):
(wgpuRenderBundleSetLabel):
* Source/WebGPU/WebGPU/RenderBundleEncoder.mm:
(WebGPU::RenderBundleEncoder::makeInvalid):
(WebGPU::RenderBundleEncoder::currentRenderCommand):
(WebGPU::RenderBundleEncoder::addResource):
(WebGPU::RenderBundleEncoder::executePreDrawCommands):
(WebGPU::RenderBundleEncoder::drawIndexed):
(WebGPU::RenderBundleEncoder::drawIndexedIndirect):
(WebGPU::RenderBundleEncoder::drawIndirect):
(WebGPU::RenderBundleEncoder::validToEncodeCommand const):
(WebGPU::RenderBundleEncoder::setBindGroup):
(WebGPU::RenderBundleEncoder::setPipeline):
(wgpuRenderBundleEncoderReference):
(wgpuRenderBundleEncoderRelease):
(wgpuRenderBundleEncoderDraw):
(wgpuRenderBundleEncoderDrawIndexed):
(wgpuRenderBundleEncoderDrawIndexedIndirect):
(wgpuRenderBundleEncoderDrawIndirect):
(wgpuRenderBundleEncoderFinish):
(wgpuRenderBundleEncoderInsertDebugMarker):
(wgpuRenderBundleEncoderPopDebugGroup):
(wgpuRenderBundleEncoderPushDebugGroup):
(wgpuRenderBundleEncoderSetBindGroupWithDynamicOffsets):
(wgpuRenderBundleEncoderSetIndexBuffer):
(wgpuRenderBundleEncoderSetPipeline):
(wgpuRenderBundleEncoderSetVertexBuffer):
(wgpuRenderBundleEncoderSetLabel):
* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::m_maxDrawCount):
(wgpuRenderPassEncoderReference):
(wgpuRenderPassEncoderRelease):
(wgpuRenderPassEncoderBeginOcclusionQuery):
(wgpuRenderPassEncoderDraw):
(wgpuRenderPassEncoderDrawIndexed):
(wgpuRenderPassEncoderDrawIndexedIndirect):
(wgpuRenderPassEncoderDrawIndirect):
(wgpuRenderPassEncoderEndOcclusionQuery):
(wgpuRenderPassEncoderEnd):
(wgpuRenderPassEncoderExecuteBundles):
(wgpuRenderPassEncoderInsertDebugMarker):
(wgpuRenderPassEncoderPopDebugGroup):
(wgpuRenderPassEncoderPushDebugGroup):
(wgpuRenderPassEncoderSetBindGroup):
(wgpuRenderPassEncoderSetBlendConstant):
(wgpuRenderPassEncoderSetIndexBuffer):
(wgpuRenderPassEncoderSetPipeline):
(wgpuRenderPassEncoderSetScissorRect):
(wgpuRenderPassEncoderSetStencilReference):
(wgpuRenderPassEncoderSetVertexBuffer):
(wgpuRenderPassEncoderSetViewport):
(wgpuRenderPassEncoderSetLabel):
* Source/WebGPU/WebGPU/RenderPipeline.mm:
(wgpuRenderPipelineReference):
(wgpuRenderPipelineRelease):
(wgpuRenderPipelineGetBindGroupLayout):
(wgpuRenderPipelineSetLabel):
* Source/WebGPU/WebGPU/Sampler.mm:
(wgpuSamplerReference):
(wgpuSamplerRelease):
(wgpuSamplerSetLabel):
* Source/WebGPU/WebGPU/ShaderModule.mm:
(WebGPU::earlyCompileShaderModule):
(wgpuShaderModuleReference):
(wgpuShaderModuleRelease):
(wgpuShaderModuleGetCompilationInfo):
(wgpuShaderModuleGetCompilationInfoWithBlock):
(wgpuShaderModuleSetLabel):
* Source/WebGPU/WebGPU/Texture.mm:
(WebGPU::imageCopyTextureSubresourceSize):
(WebGPU::Texture::errorValidatingImageCopyTexture):
(WebGPU::Texture::errorValidatingTextureCopyRange):
(wgpuTextureReference):
(wgpuTextureRelease):
(wgpuTextureCreateView):
(wgpuTextureDestroy):
(wgpuTextureSetLabel):
(wgpuTextureGetDepthOrArrayLayers):
(wgpuTextureGetDimension):
(wgpuTextureGetFormat):
(wgpuTextureGetHeight):
(wgpuTextureGetWidth):
(wgpuTextureGetMipLevelCount):
(wgpuTextureGetSampleCount):
(wgpuTextureGetUsage):
* Source/WebGPU/WebGPU/TextureView.mm:
(wgpuTextureViewReference):
(wgpuTextureViewRelease):
(wgpuTextureViewSetLabel):
* Source/WebGPU/WebGPU/XRBinding.mm:
(wgpuXRBindingReference):
(wgpuXRBindingRelease):
(wgpuBindingCreateXRProjectionLayer):
(wgpuBindingGetViewSubImage):
* Source/WebGPU/WebGPU/XRProjectionLayer.mm:
(wgpuXRProjectionLayerReference):
(wgpuXRProjectionLayerRelease):
(wgpuXRProjectionLayerStartFrame):
* Source/WebGPU/WebGPU/XRSubImage.mm:
(wgpuXRSubImageReference):
(wgpuXRSubImageRelease):
(wgpuXRSubImageGetColorTexture):
(wgpuXRSubImageGetDepthStencilTexture):
* Source/WebGPU/WebGPU/XRView.mm:
(wgpuXRViewReference):
(wgpuXRViewRelease):

Canonical link: <a href="https://commits.webkit.org/285684@main">https://commits.webkit.org/285684@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/667740a9221ceeb6e147da8a86f14b8b4370b818

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/73480 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/52909 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/26288 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/77745 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/24718 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/62040 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/694 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/57748 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16179 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/76547 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/47813 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/63213 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/38149 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/44426 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/23049 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/66233 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21049 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/79364 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/796 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/285 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/66136 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/937 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63226 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/65414 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16173 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/9264 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/7440 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/759 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/3502 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/789 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/775 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/795 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebGPU/WebGPU/APIConversions.h:
```diff
@@ -59,142 +59,142 @@ namespace WebGPU {
 // FIXME: It would be cool if we didn't have to list all these overloads, but instead could do something like bridge_cast() in WTF.
 
-inline Adapter& fromAPI(WGPUAdapter adapter)
+inline Ref<Adapter> fromAPI(WGPUAdapter adapter)
```
We don't typically do this. Instead, we instantiate Ref / RefPtr object in stack where the result of the conversion is used.

mwyrzykowski replied:
Good to know, thank you. How does that differ from the following pattern:
```
Ref<WebGPU::ObjectHeap> protectedObjectHeap() const { return m_objectHeap.get(); }
```

which seems essentially the same?

mwyrzykowski replied:
(resolving as we discussed offline)

rniwa commented on Source/WebGPU/WebGPU/Adapter.mm:
```diff
@@ -150,49 +150,49 @@
 void wgpuAdapterReference(WGPUAdapter adapter)
 {
-    WebGPU::fromAPI(adapter).ref();
+    WebGPU::protectedFromAPI(adapter)->ref();
```
This seems unnecessary. We're literally calling ref() which is an inline trivial function.

mwyrzykowski replied:
Agreed, will fix up the *Reference / *Release calls

mwyrzykowski replied:
(resolving as we discussed offline)

rniwa commented on Source/WebGPU/WebGPU/Adapter.mm:
```diff
@@ -150,49 +150,49 @@
 void wgpuAdapterRelease(WGPUAdapter adapter)
 {
-    WebGPU::fromAPI(adapter).deref();
+    WebGPU::protectedFromAPI(adapter)->deref();
```
Ditto.

rniwa commented on Source/WebGPU/WebGPU/BindGroup.mm:
```diff
@@ -924,28 +925,28 @@ static BindGroupEntryUsageData makeBindGroupEntryUsageData(BindGroupEntryUsage u
     constexpr size_t stagesPlusUndefinedCount = std::size(stagesPlusUndefined);
-    const auto& bindGroupLayout = WebGPU::fromAPI(descriptor.layout);
-    if (!bindGroupLayout.isValid() || (!bindGroupLayout.isAutoGenerated() && descriptor.entryCount != bindGroupLayout.entries().size()) || &bindGroupLayout.device() != this) {
+    Ref bindGroupLayout = WebGPU::protectedFromAPI(descriptor.layout);
```
For these code which allocates a local variable, the convention is to call non-protected variant and then store it in Ref/RefPtr.

mwyrzykowski replied:
I just made the changes in this PR but is there / will there be a style checker rule to enforce that convention? It seems easy to overlook and would otherwise pass the static analysis warning.

rniwa replied:
That's probably a good idea. Maybe we should add this to https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines as well. cc @cdumez

rniwa commented on Source/WebGPU/WebGPU/BindGroup.mm:
```diff
@@ -1400,20 +1401,20 @@ static BindGroupEntryUsageData makeBindGroupEntryUsageData(BindGroupEntryUsage u
 void wgpuBindGroupReference(WGPUBindGroup bindGroup)
 {
-    WebGPU::fromAPI(bindGroup).ref();
+    WebGPU::protectedFromAPI(bindGroup)->ref();
```
Again, this seems redundant.

rniwa commented on Source/WebGPU/WebGPU/BindGroup.mm:
```diff
@@ -1400,20 +1401,20 @@ static BindGroupEntryUsageData makeBindGroupEntryUsageData(BindGroupEntryUsage u
 void wgpuBindGroupRelease(WGPUBindGroup bindGroup)
 {
-    WebGPU::fromAPI(bindGroup).deref();
+    WebGPU::protectedFromAPI(bindGroup)->deref();
```
Ditto.

geoffreygaren commented on Source/WebGPU/WebGPU/RenderPassEncoder.mm:
```diff
@@ -1381,106 +1381,106 @@ void wgpuRenderPassEncoderRelease(WGPURenderPassEncoder renderPassEncoder)
     if (buffer)
-        optionalBuffer = &WebGPU::fromAPI(buffer);
-    WebGPU::fromAPI(renderPassEncoder).setVertexBuffer(slot, optionalBuffer, offset, size);
+        optionalBuffer = RefPtr { WebGPU::protectedFromAPI(buffer).ptr() };
```
RefPtr has a Ref<T>&& constructor, so you can drop ".ptr()" to reduce refcount churn. And that might also allow you to drop "RefPtr { }".

geoffreygaren commented on Source/WebGPU/WebGPU/RenderPipeline.mm:
```diff
@@ -1376,14 +1376,14 @@ static uint32_t componentsForDataType(MTLDataType dataType)
             return returnInvalidRenderPipeline(*this, isAsync, "Fragment has extra chain"_s);
 
-        fragmentModule = &WebGPU::fromAPI(fragmentDescriptor.module);
+        fragmentModule = RefPtr { WebGPU::protectedFromAPI(fragmentDescriptor.module).ptr() };
```
Ditto

geoffreygaren commented on Source/WebGPU/WebGPU/Queue.mm:
```diff
@@ -495,8 +495,8 @@ static void invalidateCommandBuffers(Vector<std::reference_wrapper<CommandBuffer
 
-    auto& texture = fromAPI(destination.texture);
-    if (texture.isDestroyed()) {
+    Ref texture = Ref { fromAPI(destination.texture) };
```
protectedFromAPI() would be slightly more style consistent here.

rniwa replied:
More conventional code is
```
Ref texture = fromAPI(...);
```

geoffreygaren commented on Source/WebGPU/WebGPU/Queue.mm:
```diff
@@ -531,13 +531,13 @@ static void invalidateCommandBuffers(Vector<std::reference_wrapper<CommandBuffer
     auto dataByteSize = data.size();
-    auto& texture = fromAPI(destination.texture);
-    if (texture.isDestroyed()) {
+    Ref texture = Ref { fromAPI(destination.texture) };
```
Ditto

rniwa replied:
Ditto

geoffreygaren commented on Source/WebGPU/WebGPU/RenderPassEncoder.mm:
```diff
@@ -102,22 +102,22 @@
 
-        auto& texture = fromAPI(attachment.view);
-        if (texture.isDestroyed())
+        Ref texture = Ref { fromAPI(attachment.view) };
```
Ditto.

geoffreygaren commented on Source/WebGPU/WebGPU/RenderPassEncoder.mm:
```diff
@@ -102,22 +102,22 @@
-            auto& texture = fromAPI(attachment.resolveTarget);
-            texture.setCommandEncoder(parentEncoder);
-            texture.setPreviouslyCleared();
+            Ref texture = Ref { fromAPI(attachment.resolveTarget) };
```
Ditto.

geoffreygaren commented on Source/WebGPU/WebGPU/RenderPassEncoder.mm:
```diff
@@ -131,29 +131,29 @@
-        textureView.setCommandEncoder(parentEncoder);
-        id<MTLTexture> depthTexture = textureView.isDestroyed() ? nil : textureView.texture();
-        if (textureView.width() && !m_renderTargetWidth) {
+        Ref textureView = Ref { fromAPI(attachment->view) };
```
Ditto.

geoffreygaren commented on Source/WebGPU/WebGPU/Texture.mm:
```diff
@@ -3300,21 +3300,21 @@ static WGPUExtent3D imageCopyTextureSubresourceSize(const WGPUImageCopyTexture&
         return @"imageCopyTexture is not valid";
 
-    if (imageCopyTexture.mipLevel >= fromAPI(imageCopyTexture.texture).mipLevelCount())
+    if (imageCopyTexture.mipLevel >= Ref { fromAPI(imageCopyTexture.texture) }->mipLevelCount())
```
protectedFromAPI() would make chaining clearer here.

geoffreygaren reviewed (approved):
r=me

mwyrzykowski committed:
[WebGPU] Safe C++ adaption of APIConversions.h
https://bugs.webkit.org/show_bug.cgi?id=281888
rdar://138358898

Reviewed by Geoffrey Garen.

Static analyzer indicated APIConversions resulted in warnings, correct
them all by switching the return type of fromAPI to Ref<T> from T&.

* Source/WebGPU/WebGPU/APIConversions.h:
(WebGPU::fromAPI):
* Source/WebGPU/WebGPU/Adapter.mm:
(wgpuAdapterReference):
(wgpuAdapterRelease):
(wgpuAdapterEnumerateFeatures):
(wgpuAdapterGetLimits):
(wgpuAdapterGetProperties):
(wgpuAdapterHasFeature):
(wgpuAdapterRequestDevice):
(wgpuAdapterRequestDeviceWithBlock):
(wgpuAdapterXRCompatible):
* Source/WebGPU/WebGPU/BindGroup.mm:
(WebGPU::makeBindGroupEntryUsageData):
(WebGPU::Device::createBindGroup):
(wgpuBindGroupReference):
(wgpuBindGroupRelease):
(wgpuBindGroupSetLabel):
(wgpuBindGroupUpdateExternalTextures):
* Source/WebGPU/WebGPU/BindGroupLayout.mm:
(wgpuBindGroupLayoutReference):
(wgpuBindGroupLayoutRelease):
(wgpuBindGroupLayoutSetLabel):
* Source/WebGPU/WebGPU/Buffer.mm:
(wgpuBufferReference):
(wgpuBufferRelease):
(wgpuBufferDestroy):
(wgpuBufferGetMapState):
(wgpuBufferGetMappedRange):
(wgpuBufferGetBufferContents):
(wgpuBufferGetInitialSize):
(wgpuBufferGetCurrentSize):
(wgpuBufferMapAsync):
(wgpuBufferMapAsyncWithBlock):
(wgpuBufferUnmap):
(wgpuBufferSetLabel):
(wgpuBufferGetUsage):
(wgpuBufferCopy):
* Source/WebGPU/WebGPU/CommandBuffer.mm:
(wgpuCommandBufferReference):
(wgpuCommandBufferRelease):
(wgpuCommandBufferSetLabel):
* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::errorValidatingTimestampWrites):
(WebGPU::CommandEncoder::errorValidatingRenderPassDescriptor const):
(WebGPU::CommandEncoder::beginRenderPass):
(WebGPU::CommandEncoder::errorValidatingImageCopyBuffer const):
(WebGPU::CommandEncoder::errorValidatingCopyBufferToTexture const):
(WebGPU::CommandEncoder::copyBufferToTexture):
(WebGPU::CommandEncoder::errorValidatingCopyTextureToBuffer const):
(WebGPU::CommandEncoder::clearTextureIfNeeded):
(WebGPU::CommandEncoder::copyTextureToBuffer):
(WebGPU::errorValidatingCopyTextureToTexture):
(WebGPU::CommandEncoder::copyTextureToTexture):
(wgpuCommandEncoderReference):
(wgpuCommandEncoderRelease):
(wgpuCommandEncoderBeginComputePass):
(wgpuCommandEncoderBeginRenderPass):
(wgpuCommandEncoderCopyBufferToBuffer):
(wgpuCommandEncoderCopyBufferToTexture):
(wgpuCommandEncoderCopyTextureToBuffer):
(wgpuCommandEncoderCopyTextureToTexture):
(wgpuCommandEncoderClearBuffer):
(wgpuCommandEncoderFinish):
(wgpuCommandEncoderInsertDebugMarker):
(wgpuCommandEncoderPopDebugGroup):
(wgpuCommandEncoderPushDebugGroup):
(wgpuCommandEncoderResolveQuerySet):
(wgpuCommandEncoderWriteTimestamp):
(wgpuCommandEncoderSetLabel):
* Source/WebGPU/WebGPU/ComputePassEncoder.mm:
(wgpuComputePassEncoderReference):
(wgpuComputePassEncoderRelease):
(wgpuComputePassEncoderDispatchWorkgroups):
(wgpuComputePassEncoderDispatchWorkgroupsIndirect):
(wgpuComputePassEncoderEnd):
(wgpuComputePassEncoderInsertDebugMarker):
(wgpuComputePassEncoderPopDebugGroup):
(wgpuComputePassEncoderPushDebugGroup):
(wgpuComputePassEncoderSetBindGroup):
(wgpuComputePassEncoderSetPipeline):
(wgpuComputePassEncoderSetLabel):
* Source/WebGPU/WebGPU/ComputePipeline.mm:
(wgpuComputePipelineReference):
(wgpuComputePipelineRelease):
(wgpuComputePipelineGetBindGroupLayout):
(wgpuComputePipelineSetLabel):
* Source/WebGPU/WebGPU/Device.mm:
* Source/WebGPU/WebGPU/ExternalTexture.mm:
(wgpuExternalTextureReference):
(wgpuExternalTextureRelease):
(wgpuExternalTextureDestroy):
(wgpuExternalTextureUndestroy):
(wgpuExternalTextureUpdate):
* Source/WebGPU/WebGPU/Instance.mm:
(wgpuInstanceReference):
(wgpuInstanceRelease):
(wgpuInstanceCreateSurface):
(wgpuInstanceProcessEvents):
(wgpuInstanceRequestAdapter):
(wgpuInstanceRequestAdapterWithBlock):
(wgpuBufferIsValid):
(wgpuAdapterIsValid):
(wgpuBindGroupIsValid):
(wgpuBindGroupLayoutIsValid):
(wgpuCommandBufferIsValid):
(wgpuCommandEncoderIsValid):
(wgpuComputePassEncoderIsValid):
(wgpuComputePipelineIsValid):
(wgpuDeviceIsValid):
(wgpuExternalTextureIsValid):
(wgpuPipelineLayoutIsValid):
(wgpuPresentationContextIsValid):
(wgpuQuerySetIsValid):
(wgpuQueueIsValid):
(wgpuRenderBundleEncoderIsValid):
(wgpuRenderBundleIsValid):
(wgpuRenderPassEncoderIsValid):
(wgpuRenderPipelineIsValid):
(wgpuSamplerIsValid):
(wgpuShaderModuleIsValid):
(wgpuTextureIsValid):
(wgpuTextureViewIsValid):
(wgpuXRBindingIsValid):
(wgpuXRSubImageIsValid):
(wgpuXRProjectionLayerIsValid):
(wgpuXRViewIsValid):
* Source/WebGPU/WebGPU/IsValidToUseWith.h:
(WebGPU::isValidToUseWith):
* Source/WebGPU/WebGPU/PipelineLayout.mm:
(wgpuPipelineLayoutReference):
(wgpuPipelineLayoutRelease):
(wgpuPipelineLayoutSetLabel):
* Source/WebGPU/WebGPU/PresentationContext.mm:
(wgpuSurfaceReference):
(wgpuSurfaceRelease):
(wgpuSwapChainReference):
(wgpuSwapChainRelease):
(wgpuSurfaceGetPreferredFormat):
(wgpuSwapChainGetCurrentTexture):
(wgpuSwapChainGetCurrentTextureView):
(wgpuSwapChainPresent):
(wgpuSwapChainGetTextureAsNativeImage):
* Source/WebGPU/WebGPU/QuerySet.mm:
(wgpuQuerySetReference):
(wgpuQuerySetRelease):
(wgpuQuerySetDestroy):
(wgpuQuerySetSetLabel):
(wgpuQuerySetGetCount):
(wgpuQuerySetGetType):
* Source/WebGPU/WebGPU/Queue.mm:
(WebGPU::Queue::clearTextureIfNeeded):
(WebGPU::Queue::writeTexture):
(wgpuQueueReference):
(wgpuQueueRelease):
(wgpuQueueOnSubmittedWorkDone):
(wgpuQueueOnSubmittedWorkDoneWithBlock):
(wgpuQueueSubmit):
(wgpuQueueWriteBuffer):
(wgpuQueueWriteTexture):
(wgpuQueueSetLabel):
* Source/WebGPU/WebGPU/RenderBundle.mm:
(wgpuRenderBundleReference):
(wgpuRenderBundleRelease):
(wgpuRenderBundleSetLabel):
* Source/WebGPU/WebGPU/RenderBundleEncoder.mm:
(WebGPU::RenderBundleEncoder::makeInvalid):
(WebGPU::RenderBundleEncoder::currentRenderCommand):
(WebGPU::RenderBundleEncoder::addResource):
(WebGPU::RenderBundleEncoder::executePreDrawCommands):
(WebGPU::RenderBundleEncoder::drawIndexed):
(WebGPU::RenderBundleEncoder::drawIndexedIndirect):
(WebGPU::RenderBundleEncoder::drawIndirect):
(WebGPU::RenderBundleEncoder::validToEncodeCommand const):
(WebGPU::RenderBundleEncoder::setBindGroup):
(WebGPU::RenderBundleEncoder::setPipeline):
(wgpuRenderBundleEncoderReference):
(wgpuRenderBundleEncoderRelease):
(wgpuRenderBundleEncoderDraw):
(wgpuRenderBundleEncoderDrawIndexed):
(wgpuRenderBundleEncoderDrawIndexedIndirect):
(wgpuRenderBundleEncoderDrawIndirect):
(wgpuRenderBundleEncoderFinish):
(wgpuRenderBundleEncoderInsertDebugMarker):
(wgpuRenderBundleEncoderPopDebugGroup):
(wgpuRenderBundleEncoderPushDebugGroup):
(wgpuRenderBundleEncoderSetBindGroupWithDynamicOffsets):
(wgpuRenderBundleEncoderSetIndexBuffer):
(wgpuRenderBundleEncoderSetPipeline):
(wgpuRenderBundleEncoderSetVertexBuffer):
(wgpuRenderBundleEncoderSetLabel):
* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::m_maxDrawCount):
(wgpuRenderPassEncoderReference):
(wgpuRenderPassEncoderRelease):
(wgpuRenderPassEncoderBeginOcclusionQuery):
(wgpuRenderPassEncoderDraw):
(wgpuRenderPassEncoderDrawIndexed):
(wgpuRenderPassEncoderDrawIndexedIndirect):
(wgpuRenderPassEncoderDrawIndirect):
(wgpuRenderPassEncoderEndOcclusionQuery):
(wgpuRenderPassEncoderEnd):
(wgpuRenderPassEncoderExecuteBundles):
(wgpuRenderPassEncoderInsertDebugMarker):
(wgpuRenderPassEncoderPopDebugGroup):
(wgpuRenderPassEncoderPushDebugGroup):
(wgpuRenderPassEncoderSetBindGroup):
(wgpuRenderPassEncoderSetBlendConstant):
(wgpuRenderPassEncoderSetIndexBuffer):
(wgpuRenderPassEncoderSetPipeline):
(wgpuRenderPassEncoderSetScissorRect):
(wgpuRenderPassEncoderSetStencilReference):
(wgpuRenderPassEncoderSetVertexBuffer):
(wgpuRenderPassEncoderSetViewport):
(wgpuRenderPassEncoderSetLabel):
* Source/WebGPU/WebGPU/RenderPipeline.mm:
(wgpuRenderPipelineReference):
(wgpuRenderPipelineRelease):
(wgpuRenderPipelineGetBindGroupLayout):
(wgpuRenderPipelineSetLabel):
* Source/WebGPU/WebGPU/Sampler.mm:
(wgpuSamplerReference):
(wgpuSamplerRelease):
(wgpuSamplerSetLabel):
* Source/WebGPU/WebGPU/ShaderModule.mm:
(WebGPU::earlyCompileShaderModule):
(wgpuShaderModuleReference):
(wgpuShaderModuleRelease):
(wgpuShaderModuleGetCompilationInfo):
(wgpuShaderModuleGetCompilationInfoWithBlock):
(wgpuShaderModuleSetLabel):
* Source/WebGPU/WebGPU/Texture.mm:
(WebGPU::imageCopyTextureSubresourceSize):
(WebGPU::Texture::errorValidatingImageCopyTexture):
(WebGPU::Texture::errorValidatingTextureCopyRange):
(wgpuTextureReference):
(wgpuTextureRelease):
(wgpuTextureCreateView):
(wgpuTextureDestroy):
(wgpuTextureSetLabel):
(wgpuTextureGetDepthOrArrayLayers):
(wgpuTextureGetDimension):
(wgpuTextureGetFormat):
(wgpuTextureGetHeight):
(wgpuTextureGetWidth):
(wgpuTextureGetMipLevelCount):
(wgpuTextureGetSampleCount):
(wgpuTextureGetUsage):
* Source/WebGPU/WebGPU/TextureView.mm:
(wgpuTextureViewReference):
(wgpuTextureViewRelease):
(wgpuTextureViewSetLabel):
* Source/WebGPU/WebGPU/XRBinding.mm:
(wgpuXRBindingReference):
(wgpuXRBindingRelease):
(wgpuBindingCreateXRProjectionLayer):
(wgpuBindingGetViewSubImage):
* Source/WebGPU/WebGPU/XRProjectionLayer.mm:
(wgpuXRProjectionLayerReference):
(wgpuXRProjectionLayerRelease):
(wgpuXRProjectionLayerStartFrame):
* Source/WebGPU/WebGPU/XRSubImage.mm:
(wgpuXRSubImageReference):
(wgpuXRSubImageRelease):
(wgpuXRSubImageGetColorTexture):
(wgpuXRSubImageGetDepthStencilTexture):
* Source/WebGPU/WebGPU/XRView.mm:
(wgpuXRViewReference):
(wgpuXRViewRelease):

Canonical link: https://commits.webkit.org/285684@main
